app/src/ts.app.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * Broadcast Message to one or more apps.
+ * @param {array|string} appIds List of apps to receive the message, supports glob
+ * Examples:
+ * * ['Tradeshift.Developer', 'Tradeshift.DeveloperDemo'] (matches Tradeshift.Developer and Tradeshift.DeveloperDemo)
+ * * 'Tradeshift.Developer' (matches Tradeshift.Developer)
+ * * '*' (matches any app)
+ * * 'Tradeshift.*' (matches all Tradeshift apps)
+ * * 'Tradeshift.??Y' (matches Tradeshift.Buy, Tradeshift.Pay, etc.)
+ * @param {string} key The key/subject of the event
+ * @param {object} data Data to be sent with the event, has to support the structured clone algorithm
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
+ */
+export function broadcast(appIds, key, data) {
+	const content = stringify(appIds, key, data);
+	parent.postMessage(content, '*');
+}
+
+/**
+ * Subscribe to messages from one or more apps.
+ * @param {array|string} appIds List of apps to receive the message, supports glob
+ * @return {Listener} Listener object
+ */
+export function subscribe(appIds) {
+	return new Listener(appIds);
+}
+
+const BROADCAST_PREFIX = 'app-broadcast:';
+
+/**
+ * Message listener for one or more apps.
+ */
+class Listener {
+	/**
+	 * Create the listener.
+	 * @param {array|string} appIds 
+	 */
+	constructor(appIds) {
+		this.appIds = appIds;
+		this.events = new Map();
+	}
+
+	/**
+	 * Listen to message and call handler.
+	 * @param {string} key The key/subject of the event.
+	 * @param {function} callback The handler of the event.
+	 * @return {Listener} Listener object
+	 */
+	on(key, callback) {
+		if (!key || !callback) {
+			console.warn(`We need two parameters key and callback`);
+			return this;
+		}
+		if (this.events.has(key)) {
+			console.warn(`You have listened the message of ${key}`);
+			return this;
+		}
```

Comment:
But what if the callback is different now? I guess we should only return if the callback is the same; and even so, I can't see a reason to log a warning. Also, the way this is set up, I assume that we should `off` the old key first before we register the new callback (with the same key).